### ISSUE TITLE
composite,amsr-method() is broken

### ISSUE THREAD

**Issue by dankelley** [dankelley/oce#2147]
The AMSR file format has changed, and I suspect that's why `composite` is now failing, as shown below.  I'll look into this.

<img width="1049" alt="Screenshot 2023-09-08 at 4 40 44 PM" src="https://github.com/dankelley/oce/assets/99469/d1b35ab8-7d29-4cb2-9ac5-7398e585971b">

**Comment by dankelley** [comment]
As expected, it's because of the file-format change. Things used to be type `raw` and now they are type `numeric`.  (This also explains the large increase in file size.  That, and the fact that now the N masks are each stored as matrices, instead of one matrix with N values.)

Below is one place where I'll make a fix, and likely also in the `do_` code (which is in C++, for speed).

<img width="809" alt="Screenshot 2023-09-08 at 5 05 40 PM" src="https://github.com/dankelley/oce/assets/99469/c8b668f9-1cfc-4796-9064-9aea0198ab43">

**Comment by dankelley** [comment]
In the old format, things like SST were stored as `raw` entries, and special values were set aside for the various masks.  Dealing with them in combining objects was slow in R, so it was done in C++.  But, now, the masks are stored as matrices, so I can make boolean operations to turn the SST (etc) values into NAs.  And then I can make a 3D array by stacking those arrays, and use `apply()` with `mean` as the function.  In this scheme, no loops are required, and R will be doing the grunt work in C, so we ought to gain much simpler programming -- in R, which users can be anticipated to know -- and I don't think it will be much slower.

I'll think this through a bit and try coding it later today.

**Comment by dankelley** [comment]
Here's a test of creating a mask of bad values. Simple R, and 30X faster than the blink of an eye, so no problem for the user.
```
> system.time(bad <- with(d1@data, coastMask | landMask | noObsMask | seaIceMask))
   user  system elapsed 
  0.015   0.001   0.016 
```

**Comment by dankelley** [comment]
Hm.  I'm making a 3D array, filling in e.g. `A[,,1]` with `object`, `A[,,2]` with `...[[1]` etc but my test with just 3 amsr objects takes over a minute.  The time is soaked up in `apply()`, I think.

Maybe I'll need to use C++ after all, because looking through things is very fast there.

**Comment by dankelley** [comment]
I think it's time for me to bite the bullet and learn RcppArmadillo, because it apparently has nice support for 3D arrays.  (I don't think Rcpp has such good support past 2D.  I may be wrong on that, but there are a *lot* of packages using RcppArmadillo.  My guess is that any new C++ work will be done there, not in simple C++ and not in Rcpp.  So it's time.  Maybe.  I need to do some reading to tell.  (It's not so much that this will speed up this particular task, because I am comfortable in C, but I don't think most R users are, and RcppArmadillo is at a much higher-level so it ought to be at least somewhat readible to many users.)

Eddelbuettel, Dirk, and Conrad Sanderson. “RcppArmadillo: Accelerating R with High-Performance C++ Linear Algebra.” Computational Statistics & Data Analysis 71 (March 2014): 1054–63. https://doi.org/10.1016/j.csda.2013.02.005.

**Comment by dankelley** [comment]
OK, I know this makes me look lazy, but I found it too hard to locate the information I needed on RcppArmadillo and I realized that I could just do this in R, computing the mean value simply by adding etc.

Now, the combination of 3 objects takes 0.6 seconds.  Not bad compared with the 70 or so seconds I had using the `apply` method, and a *lot* easier than trying to figure out the best way with RcppArmadillo.
